pyro/infer/mcmc/mh.py

```diff
@@ -0,0 +1,255 @@
+from __future__ import absolute_import, division, print_function
+
+import pyro
+import pyro.distributions as dist
+from pyro import poutine
+from pyro.infer.mcmc.trace_kernel import TraceKernel
+from pyro.util import ng_ones, ng_zeros
+from torch import Tensor as T
+from torch.autograd import Variable as V
+from numpy.random import choice
+from numpy import isfinite
+from torch import log as tlog
+
+_R = '_RETURN'
+
+
+def VT(val):
+    return V(T(val))
+
+
+def VTA(val):
+    return VT([val])
+
+
+def assert_valid_trace(trace):
+    assert not any(['fn' in n and isinstance(n['fn'], pyro._Subsample)
+                    for nid, n in trace.nodes(data=True)]), \
+                    "Single Site MH does not currently handle mapdata in model."
+
+
+# get all the sample sites
+def sample_sites(trace, minus=set()):
+    return [nid for nid, n in trace.nodes(data=True)
+            if nid not in minus and n["type"] == 'sample'
+            and not n["is_observed"]]
+
+
+class NormalProposal():
+    def __init__(self, mu, sigma, tune_frequency=100):
+        self.mu = mu
+        self.sigma = sigma
+        self.scale = 1.
+        self.tune_frequency = tune_frequency
+        self._tune_cnt = self.tune_frequency
+
+    def sample(self, site_node):
+        cur_val = site_node['value']
+        mu = self.mu.expand_as(cur_val)
+        sigma = self.scale*self.sigma
+
+        # sigma just a number? expand to equal the must
+        if tuple(sigma.data.shape) != mu.data.shape:
+            sigma = sigma.expand_as(mu.data)
+
+        return cur_val + dist.Normal(mu, sigma).sample()
+
+    def log_pdf_given(self, y, given_x):
+        # center our distribution at mu = given_x, then
+        # get log pdf of y
+        mu = given_x
+        sigma = self.scale*self.sigma
+        return dist.Normal(mu, sigma).log_pdf(y)
+
+    # borrowed from:
```

Comment:
We need to either paste of their licenses here or find an original source to cite and rewrite the code.